### ISSUE TITLE
Four GRPO training-stability bugs found during Qwen3.5-9B LoRA runs (reproducible, paired with #49)

### ISSUE THREAD

**Issue by SudarkinV** [Goekdeniz-Guelmez/mlx-lm-lora#50]
# Four GRPO training-stability bugs found during Qwen3.5-9B LoRA runs

**Package**: `mlx-lm-lora`, file `mlx_lm_lora/trainer/grpo_trainer.py`
**Hardware**: MacBook Pro M4 Max, 36 GB unified memory, MLX 0.31
**Use case**: GRPO LoRA fine-tuning of Qwen3.5-9B-MLX-4bit on a
structured-field extraction task with a composite JSON-quality reward
(registered via `@register_reward_function()`).
**Paired PR**: #49 (please review after confirming the bugs below)

This is the combined report for four distinct bugs I hit in a single
GRPO run. They can be split into separate issues if you prefer —
let me know.

---

## TL;DR

| # | Bug | Symptom | Root cause |
|---|-----|---------|------------|
| 1 | `importance_sampling_level=None` default | Training runs, but **no policy-gradient signal**; only KL penalty moves weights | `None` falls through to the `zeros_like` branch, making `coef_1 = exp(0) = 1` — a constant w.r.t. the policy |
| 2 | Advantage explosion with small `GROUP_SIZE` | `NaN` gradients ~200–400 iterations in | Empirical per-group std can approach 0 for `G=2`; unnormalised advantages blow up |
| 3 | `@mx.compile` + LoRA | Model **degrades** 90.2% → 85.4% over 20 iterations (smoking gun below) | `mx.compile` captures the LoRA adapter weights as constants in the compiled graph; `nn.value_and_grad` returns zero gradient for them |
| 4 | Undocumented `importance_sampling_level='none'` | Easy to set by mistake; silently trains only through KL penalty | Docstring did not flag the implication |

---

## Setup (fully public, reproducible)

```yaml
# grpo_config.yaml
model: mlx-community/Qwen3.5-9B-MLX-4bit
train: true
fine_tune_type: lora
algorithm: grpo

batch_size: 1
group_size: 2
max_tokens: 1536
learning_rate: 1e-5
iters: 100
steps_per_eval: 25

num_layers: 16
lora_parameters:
  rank: 8
  scale: 2.0
  dropout: 0.0
  keys: [self_attn.q_proj, self_attn.v_proj]

# custom reward function registered via
# @register_reward_function decorator
reward_functions:
  - composite_json_quality
```

The reward function returns a scalar in approximately `[-1, +1]`
built from standard JSON-extraction components (schema validity,
enum membership, fuzzy field-level match, hallucination penalty).
The specific reward does not matter for bug reproduction — any reward
function that produces non-trivial group variance shows the same
behaviour.

---

### Location

`mlx_lm_lora/trainer/grpo_trainer.py`, around line 432–440:

```python
if importance_sampling_level == "token":
    log_importance_weights = per_token_logps - old_per_token_logps
elif importance_sampling_level == "sequence":
    ...
elif importance_sampling_level is None or importance_sampling_level == "none":
    log_importance_weights = mx.zeros_like(per_token_logps)   # ← HERE
```

### What happens

When `importance_sampling_level` is the default (`None`), the
`zeros_like` branch fires, so

```python
coef_1 = mx.exp(log_importance_weights) = mx.exp(0) = 1
```

`coef_1` becomes a constant `1` with no dependency on the policy
parameters. The resulting loss has **no gradient flowing through the
policy network at all**; only the KL-penalty term (if `beta > 0`)
produces any update.

### Minimal demonstration

```python
import mlx.core as mx
import mlx.nn as nn

per_token_logps = mx.array([[-0.5, -1.0, -0.3]], dtype=mx.float32)
old_per_token_logps = mx.array([[-0.6, -0.9, -0.4]], dtype=mx.float32)

# Default (None) path:
log_w_bug  = mx.zeros_like(per_token_logps)
coef1_bug  = mx.exp(log_w_bug)          # [[1, 1, 1]]  — constant

# Intended ("token") path:
log_w_ok   = per_token_logps - old_per_token_logps
coef1_ok   = mx.exp(log_w_ok)           # [[1.10, 0.90, 1.10]]

print("bug coef_1 =", coef1_bug)   # does NOT depend on policy logps
print("ok  coef_1 =", coef1_ok)    # does depend on policy logps
```

### Fix

Default to `"token"` (the standard GRPO token-level ratio π_θ/π_old),
which is already the intended semantic when the user does not
explicitly disable importance sampling.

---

### Location

`mlx_lm_lora/trainer/grpo_trainer.py`, around line 280–290:

```python
for j, idx in enumerate(group_indices):
    advantages[idx] = (prompt_rewards[j] - mean_reward) / (
        std_reward + 1e-8
    )
```

### What happens

On `M4 Max` with `GROUP_SIZE=2` and `max_tokens=1536` (common
concurrent-memory sweet spot for 9B LoRA), the empirical std can be
essentially zero when both completions produce identical rewards. The
advantage becomes `(r − r̄) / 1e-8` which saturates to ±1e8, gradients
explode, and `NaN` appears a few hundred iterations in.

### Reproduction

```python
import mlx.core as mx

# Two near-identical rewards in a group
rewards = mx.array([0.7, 0.7 + 1e-7])
mean = rewards.mean()
std  = rewards.std()
advantages = (rewards - mean) / (std + 1e-8)
print("std =", std.item(), "advantages =", advantages)
# std ≈ 5e-8 → advantages saturate to huge values
```

### Fix

```python
adv = (prompt_rewards[j] - mean_reward) / (std_reward + 1e-8)
advantages[idx] = mx.clip(adv, -2.0, 2.0)
```

`±2.0` is a standard choice that still lets GRPO discriminate winners
from losers but prevents rare near-identical groups from blowing up
the gradient.

---

## Bug 3 — `@mx.compile` silently drops LoRA gradients (the big one)

This is the bug that made model quality *degrade* during GRPO training.

### Location

`mlx_lm_lora/trainer/grpo_trainer.py`, around line 780:

```python
@partial(mx.compile, inputs=state, outputs=state)
def step(...):
    loss_and_grad_fn = nn.value_and_grad(model, loss_fn)
    (lvalue, metrics), grads = loss_and_grad_fn(...)
    ...
```

### What happens

`mx.compile` traces the function once and captures any array that is
not in the `inputs` tree as a *graph constant*. LoRA adapter arrays
(`lora_A`, `lora_B` registered on the wrapped model) are attached to
sub-modules and therefore **do** count as such constants. When
`nn.value_and_grad` then differentiates through the compiled graph,
the gradient w.r.t. those constants is zero by definition.

Full-parameter training is not affected, because every trainable
weight is already referenced through `state` and correctly traced.
Only LoRA (or any other adapter technique that injects new
trainable tensors after `state` is built) hits this silent zero-grad
condition.

### Smoking gun: real eval numbers from my training run

Eval set is a held-out validation split (150 samples, structured-field
extraction). Total score is a weighted average of 7 field-level
metrics, higher is better. Field names below are generic placeholders
for the task's target fields.

| Checkpoint                 | Total   | field_A | field_B | field_C |
|----------------------------|---------|---------|---------|---------|
| SFT baseline (pre-GRPO)    | **90.2%** | 97.5%   | 88.1%   | 81.7%   |
| After GRPO iter 5          | 90.1%   | 95.6%   | 88.3%   | 83.4%   |
| After GRPO iter 10         | 87.6%   | 93.1%   | 88.3%   | 78.2%   |
| After GRPO iter 15         | 86.8%   | 93.1%   | 88.1%   | 78.4%   |
| After GRPO iter 20         | **85.4%** | 93.1%   | 91.3%   | **65.2%** |

Quality **degrades** from 90.2% → 85.4% over 20 iterations — and
`field_C`, one of the reward-relevant fields, collapses from 81.7%
→ 65.2%. This is the opposite of what GRPO is supposed to do when
its reward function scores exactly these fields. Once `mx.compile`
is disabled on the step function, gradients become non-zero again
and the metrics stop regressing.

### Reproduction

```python
import mlx.core as mx
import mlx.nn as nn
from functools import partial

class Adapter(nn.Module):
    def __init__(self):
        super().__init__()
        self.lora_A = mx.zeros((4, 4))   # simulates LoRA weight

    def __call__(self, x):
        return x @ self.lora_A

model = Adapter()
state = [model.lora_A]   # this is what the GRPO trainer does

def loss_fn(model, x):
    return (model(x) ** 2).sum()

# Without @mx.compile — gradients correct.
g1 = nn.value_and_grad(model, loss_fn)(model, mx.ones((1, 4)))[1]
print("no-compile grad norm:", sum((g**2).sum() for g in g1.values()).item())

# With @mx.compile — gradients zero.
@partial(mx.compile, inputs=state, outputs=state)
def step(x):
    return nn.value_and_grad(model, loss_fn)(model, x)[1]

g2 = step(mx.ones((1, 4)))
print("compiled grad norm:", sum((g**2).sum() for g in g2.values()).item())
# prints 0.0
```

### Fix

Disable `@mx.compile` on the GRPO step function when LoRA/DoRA/any
adapter is active, i.e. when `fine_tune_type != "full"`. Regular
full-parameter GRPO is unaffected and keeps its compile benefit.

---

## Bug 4 — docstring warning for `importance_sampling_level='none'`

Minor, but users (me included) hit this. Explicit `'none'` turns off
the policy-gradient term and makes KL penalty the only gradient
source. Added a warning to the docstring so this is a conscious
choice rather than a silent footgun.

---

## Environment

```
MLX             0.31.1
mlx-lm          0.30.7
mlx-lm-lora     0.4.5 (plus PR #49)
Python          3.12.13
macOS           15.6.1, M4 Max 36 GB
```

---

## What I'd appreciate from maintainers

1. Confirmation that the four bugs reproduce on your side, or hints
   if some of them are in fact intended behaviour I am
   misunderstanding.
2. Feedback on PR #49 — happy to split it into four smaller PRs if
   that makes review easier, or to add regression tests.
3. Any guidance on the `@mx.compile` fix — I chose to disable compile
   on the LoRA path specifically; an alternative is to lift adapter
   weights into the `state` tree but that touches the training loop
   more invasively, so I went with the minimal change.

Happy to share sanitised training-log excerpts and the raw eval JSON
privately if that helps — didn't attach them here to keep the issue
self-contained and to avoid pasting dataset-specific content into
a public report.

Thanks for maintaining `mlx-lm-lora` — this project has been a huge
help in getting serious LoRA fine-tuning working on M-series Macs.

**Comment by Goekdeniz-Guelmez** [comment]
hey @SudarkinV 

thanks for flagging this!

However these are not really code level bugs. The sampling level is set to none to replicate the OG paper, you can set the sampling level in the yaml if you need it via `importance_sampling_level`.

The actual epsilon in the advantages normalization is 1e-4, not 1e-8.

model.state in input=state is a live reference that includes LoRA adapter arrays. This is identical to the pattern used in sft trainer, which works correctly for LoRA. MLX doesn't silently zero gradients for arrays listed in inputs.

**Comment by SudarkinV** [comment]
Thanks for the detailed response, @Goekdeniz-Guelmez — I owe a full apology. I re-checked everything and you're right on all three counts:

1. **`importance_sampling_level=None`** — I missed that this intentionally replicates the original GRPO paper (pure policy gradient + KL, no ratio term). The yaml field lets users opt in to token/sequence level when desired. Not a bug, a design choice I should have verified before filing.

2. **Epsilon** — I misread the code. Upstream uses `1e-4` in every version I checked, not `1e-8`. My mistake, there is no numerical explosion issue.

3. **`@mx.compile` + LoRA** — my minimal reproduction in the issue was incorrectly constructed (output signature mismatch). Looking at the actual `state = [model.state, optimizer.state, mx.random.state]` pattern, and given that the SFT trainer uses identical compile semantics and trains LoRA correctly (I use it myself), your explanation is the right one. MLX traces arrays listed in \`inputs=state\` as live references, so gradients flow through LoRA normally.

The model-quality regression I observed (90.2% → 85.4% over 20 GRPO iterations on our dataset) was real, but the root cause is almost certainly on our side — reward shaping, learning rate, or the inherent instability of GROUP_SIZE=2 on our specific data. I jumped to \"first plausible framework-level bug\" instead of eliminating configuration alternatives first.

Closing this issue and PR #49. Thanks for the patience and the clear explanation — genuinely appreciated.

**Comment by SudarkinV** [comment]
Closing — all three claims were my mistake. Apologies for the noise.

**Comment by Goekdeniz-Guelmez** [comment]
no worries!
